### DISCUSSION TITLE
Brew not found after installation

### DISCUSSION THREAD

**Discussion by Motzumoto** [Homebrew/discussions#4194] [Getting started] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 3.6.21
ORIGIN: https://github.com/Homebrew/brew
HEAD: 932d2cf3b77c9439a57b6a43577fc8d3b6399a62
Last commit: 3 days ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 0047c6ead1f8e6aa0c78c4aedd7bacd62b0f3538
Core tap last commit: 2 hours ago
Core tap branch: master
Core tap JSON: 11 Feb 03:19 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0.0
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 2.6.8 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: dodeca-core 64-bit zen
Clang: 14.0.6
Git: 2.39.1 => /bin/git
Curl: 7.87.0 => /bin/curl
Kernel: Linux 6.0.0-kali6-amd64 x86_64 GNU/Linux
OS: Kali GNU/Linux Rolling (kali-rolling)
Host glibc: 2.36
/usr/bin/gcc: 12.2.0
/usr/bin/ruby: 3.1.2
glibc: N/A
gcc@11: N/A
gcc: 12.2.0
xorg: N/A
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

I install brew, i run the commands to install it, 
```
  342  /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
  343  (echo; echo 'eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"') >> /home/lain/.zprofile\n
  344  eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"\n
  345  brew
  346  brew doctor
```

and as soon as i either open a new terminal or close the one i used to install brew, it doesn't recognize the command anymore.

I'm on Kali Linux. I'm not sure what information i need to provide, so please ask and I shall provide.

The reason why this isnt #775 is because they're on a mac and i'm not. Please do not close this as its not a duplicate. Again, i run the "next steps" commands and i still run into this issue.

**Comment by gromgit** [comment, marked as answer]
Your copy-n-paste went very wrong. Assuming the output you posted was from your Zsh history:
>```
>(echo; echo 'eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"') >> /home/lain/.zprofile\n
>```
This adds the `eval` line to `/home/lain/.zprofile$'\n'` (i.e. with a extra newline at the end of the filename). This is why:
> as soon as i either open a new terminal or close the one i used to install brew, it doesn't recognize the command anymore

because the correct filename should be just `/home/lain/.zprofile`.
>```
>eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"\n
>```
This `eval`s the `shellenv` output followed by an extra newline, which fortunately has no negative impact in this situation.

To be clear, the correct instructions are:
```
(echo; echo 'eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"') >> /home/lain/.zprofile
eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"
```
You might also want to do the following:
```shell
# See what else was mistakenly added to ".zprofile-with-a-newline"
cat /home/lain/.zprofile?
# Add everything in that file to your .zprofile, then...
rm /home/lain/.zprofile?
```

**Reply by Motzumoto** [reply]
This still doesnt allow brew to be persistent for some reason
![image](https://user-images.githubusercontent.com/45925152/218241697-b3a7517c-fe85-4c6c-98c7-1409c3867b48.png)
See the screenshot linked

**Reply by gromgit** [reply]
Forgot to mention a second step: *log out and log back in again*. `.zprofile` is only read on login shells.

For the gory details on the various Zsh startup and shutdown files, read the [STARTUP/SHUTDOWN FILES section of the `zsh` man page](https://www.mankier.com/1/zsh#Startup/Shutdown_Files). If you're wondering how that answers the question "so which file do I put a specific command in?",[^1] see [here](https://unix.stackexchange.com/a/71258/98480) for a quick summary.

[^1]: Manpages tend to be incredibly dense, but for stuff like `bash` and `zsh`, the devil's in the details, so spending time to read and ponder the text, and conducting experiments to confirm your understanding, is well worth it.

**Reply by Motzumoto** [reply]
Even after logging back in brew still refuses to say persistent. 

I'm going to just wipe the entire install of brew and going to redo it, I don't know what else to do at this point.

After completely wiping the brew install and all its folders, reinstalling and entirely restarting my system, brew is still not persistent. This is so incredibly frustrating

**Reply by gromgit** [reply]
What's the output of the following commands _without manually running `brew shellenv`_?
```
getent passwd lain
echo $SHELL
echo $PATH
```

**Reply by Motzumoto** [reply]
> What's the output of the following commands _without manually running `brew shellenv`_?
```
lain:x:1000:1000:Lain,,,:/home/lain:/usr/bin/zsh
/usr/bin/zsh
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/games:/usr/games
```
this looks like an issue LOL

**Reply by gromgit** [reply]
Check to see if:
1. you have a `~/.zshrc`, and
2. you change your `PATH` in `~/.zshrc` (or another file sourced by `~/.zshrc`) to something that doesn't reference its previous value.

Also, just to be sure, what's the output of `echo ${ZDOTDIR:-"ZDOTDIR not set"}`?

**Reply by Motzumoto** [reply]
Do you need the content of `~/.zshrc`? Its quite extensive. Also the output of `echo ${ZDOTDIR:-"ZDOTDIR not set"}` is `ZDOTDIR not set`

**Reply by gromgit** [reply]
> Do you need the content of `~/.zshrc`?

No, just figured out a better way. What's the output of `zsh -xlc true 2>&1 | grep -w PATH`?

**Reply by gromgit** [reply]
> What's the output of `zsh -xlc true 2>&1 | grep -w PATH`?

Sorry, that invokes a non-interactive shell. What's the output of just `zsh -xl`?

**Reply by Motzumoto** [reply]
Its a lot, ive pasted it all into this text file: 
[output.txt](https://github.com/Homebrew/discussions/files/10717335/output.txt)

**Reply by gromgit** [reply]
Here's your problem:
```shell
$ grep -F /home/lain/.zprofile output.txt 

# Wait, why isn't ~/.zprofile being sourced?
```
It turns out your setup does this:
```shell
└─$ zsh -xl              
+/etc/zsh/zshenv:15> [[ -z /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/games:/usr/games || /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/games:/usr/games == /bin:/usr/bin ]]
+/etc/zsh/zprofile:9> emulate sh -c '. /etc/profile'
+/etc/zsh/zprofile:9> . /etc/profile
[...]
```
For your setup, it'll make more sense to add the `brew shellenv` line to `~/.zshrc` instead, which _is_ sourced:
```shell
$ grep -F /home/lain/.zshrc output.txt | wc -l
137
```

**Reply by Motzumoto** [reply]
Okay sweet, what do i have to do to fix this? Now that you've identified the problem how do we fix it

**Reply by carlocab** [reply]
```
(echo; echo 'eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"') >> /home/lain/.zshrc
```

**Reply by Motzumoto** [reply]
> ```
> (echo; echo 'eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"') >> /home/lain/.zshrc
> ```

I'm going to kiss you on the mouth (thank you)
